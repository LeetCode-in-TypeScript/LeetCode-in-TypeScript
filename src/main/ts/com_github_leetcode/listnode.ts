class ListNode {
    public val: number | null
    public next: ListNode | null

    constructor(val?: number | null, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val
        this.next = next === undefined ? null : next
    }
}

export { ListNode }
