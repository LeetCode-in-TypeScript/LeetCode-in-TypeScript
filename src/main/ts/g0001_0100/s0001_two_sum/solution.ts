// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Array #Hash_Table
// #Data_Structure_I_Day_2_Array #Level_1_Day_13_Hashmap #Udemy_Arrays #Top_Interview_150_Hashmap
// #Big_O_Time_O(n)_Space_O(n) #AI_can_be_used_to_solve_the_task
// #2025_03_22_Time_1_ms_(89.70%)_Space_59.39_MB_(8.75%)

function twoSum(nums: number[], target: number): number[] {
    const indexMap: Map<number, number> = new Map()
    for (let i = 0; i < nums.length; i++) {
        const requiredNum: number = target - nums[i]
        if (indexMap.has(requiredNum)) {
            return [indexMap.get(requiredNum)!, i]
        }
        indexMap.set(nums[i], i)
    }
    return [-1, -1]
}

export { twoSum }
