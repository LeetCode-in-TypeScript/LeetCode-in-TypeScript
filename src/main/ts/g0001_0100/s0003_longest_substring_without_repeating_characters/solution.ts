// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Sliding_Window
// #Algorithm_I_Day_6_Sliding_Window #Level_2_Day_14_Sliding_Window/Two_Pointer #Udemy_Strings
// #Big_O_Time_O(n)_Space_O(1) #2023_09_28_Time_65_ms_(90.77%)_Space_44.9_MB_(95.29%)

function lengthOfLongestSubstring(s: string): number {
    const lastIndices: number[] = new Array(256).fill(-1)
    let maxLen: number = 0
    let curLen: number = 0
    let start: number = 0
    for (let i = 0; i < s.length; i++) {
        const cur: string = s.charAt(i)
        const charCode: number = cur.charCodeAt(0)
        if (lastIndices[charCode] < start) {
            lastIndices[charCode] = i
            curLen++
        } else {
            const lastIndex: number = lastIndices[charCode]
            start = lastIndex + 1
            curLen = i - start + 1
            lastIndices[charCode] = i
        }
        if (curLen > maxLen) {
            maxLen = curLen
        }
    }
    return maxLen
}

export { lengthOfLongestSubstring }
