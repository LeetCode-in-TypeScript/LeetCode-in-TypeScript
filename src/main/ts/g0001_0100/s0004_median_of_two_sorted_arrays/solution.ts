// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Array #Binary_Search #Divide_and_Conquer
// #Big_O_Time_O(log(min(N,M)))_Space_O(1) #2023_08_26_Time_86_ms_(92.15%)_Space_48.3_MB_(59.73%)

function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
    const nums = nums1.concat(nums2)
    nums.sort((a, b) => a - b)
    const len = nums.length
    if (len % 2 === 0) {
        return (nums[len / 2 - 1] + nums[len / 2]) / 2
    } else {
        return nums[Math.floor(len / 2)]
    }
}

export { findMedianSortedArrays }
