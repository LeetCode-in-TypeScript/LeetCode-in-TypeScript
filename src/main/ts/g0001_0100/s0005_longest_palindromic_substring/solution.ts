// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Dynamic_Programming
// #Data_Structure_II_Day_9_String #Algorithm_II_Day_14_Dynamic_Programming
// #Dynamic_Programming_I_Day_17 #Udemy_Strings #Top_Interview_150_Multidimensional_DP
// #Big_O_Time_O(n)_Space_O(n) #2025_03_22_Time_8_ms_(99.14%)_Space_59.13_MB_(33.28%)

function longestPalindrome(s: string): string {
    const newStr: string[] = new Array(s.length * 2 + 1)
    newStr[0] = '#'
    for (let i = 0; i < s.length; i++) {
        newStr[2 * i + 1] = s.charAt(i)
        newStr[2 * i + 2] = '#'
    }
    const dp: number[] = new Array(newStr.length)
    let friendCenter: number = 0
    let friendRadius: number = 0
    let lpsCenter: number = 0
    let lpsRadius: number = 0
    for (let i = 0; i < newStr.length; i++) {
        dp[i] =
            friendCenter + friendRadius > i ? Math.min(dp[2 * friendCenter - i], friendCenter + friendRadius - i) : 1
        while (i + dp[i] < newStr.length && i - dp[i] >= 0 && newStr[i + dp[i]] === newStr[i - dp[i]]) {
            dp[i]++
        }
        if (friendCenter + friendRadius < i + dp[i]) {
            friendCenter = i
            friendRadius = dp[i]
        }
        if (lpsRadius < dp[i]) {
            lpsCenter = i
            lpsRadius = dp[i]
        }
    }
    return s.substring((lpsCenter - lpsRadius + 1) / 2, (lpsCenter + lpsRadius - 1) / 2)
}

export { longestPalindrome }
