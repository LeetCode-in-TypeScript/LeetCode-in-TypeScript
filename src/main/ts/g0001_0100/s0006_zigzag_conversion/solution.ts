// #Medium #String #Top_Interview_150_Array/String
// #2025_03_22_Time_2_ms_(99.08%)_Space_60.88_MB_(31.04%)

function convert(s: string, numRows: number): string {
    if (numRows < 2) {
        return s
    }
    const width = numRows * 2 - 2
    let result: string = ''
    let i
    let j
    for (let row = 0; row < numRows; row++) {
        i = row
        j = width - 2 * row
        while (i < s.length) {
            if (j > 0) {
                result += s[i]
            }
            i += j
            j = width - j
        }
    }
    return result
}

export { convert }
