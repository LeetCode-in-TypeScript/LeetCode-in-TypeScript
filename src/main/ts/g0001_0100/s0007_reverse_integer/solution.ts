// #Medium #Top_Interview_Questions #Math #Udemy_Integers
// #2023_08_25_Time_56_ms_(95.68%)_Space_43.8_MB_(99.70%)

function reverse(x: number): number {
    let remaining = Math.abs(x)
    let power = Math.floor(Math.log10(remaining))
    let reversed = 0
    while (remaining > 0) {
        const digit = remaining % 10
        remaining = (remaining - digit) / 10
        reversed += digit * 10 ** power--
        if (reversed > 2 ** 31 - 1) return 0
    }
    return x > 0 ? reversed : -reversed
}

export { reverse }
