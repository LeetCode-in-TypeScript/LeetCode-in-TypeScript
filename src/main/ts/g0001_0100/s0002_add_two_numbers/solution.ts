// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Math #Linked_List #Recursion
// #Data_Structure_II_Day_10_Linked_List #Programming_Skills_II_Day_15
// #Top_Interview_150_Linked_List #Big_O_Time_O(max(N,M))_Space_O(max(N,M))
// #AI_can_be_used_to_solve_the_task #2025_03_22_Time_2_ms_(95.82%)_Space_62.80_MB_(21.07%)

import { ListNode } from '../../com_github_leetcode/listnode'

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const dummyHead: ListNode = new ListNode(0)
    let p: ListNode | null = l1
    let q: ListNode | null = l2
    let curr: ListNode | null = dummyHead
    let carry: number = 0
    while (p !== null || q !== null) {
        const x: number = p !== null ? p.val : 0
        const y: number = q !== null ? q.val : 0
        const sum: number = carry + x + y
        carry = Math.floor(sum / 10)
        curr.next = new ListNode(sum % 10)
        curr = curr.next
        if (p !== null) {
            p = p.next
        }
        if (q !== null) {
            q = q.next
        }
    }
    if (carry > 0) {
        curr.next = new ListNode(carry)
    }
    return dummyHead.next
}

export { addTwoNumbers }
