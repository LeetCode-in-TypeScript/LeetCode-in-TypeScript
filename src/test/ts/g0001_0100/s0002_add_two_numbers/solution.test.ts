// tslint:disable:no-magic-numbers
import { ListNode } from 'src/main/ts/com_github_leetcode/listnode'
import { addTwoNumbers } from 'src/main/ts/g0001_0100/s0002_add_two_numbers/solution'
import { expect, test } from 'vitest'

test('addTwoNumbers', () => {
    expect(
        addTwoNumbers(
            new ListNode(2, new ListNode(4, new ListNode(3))),
            new ListNode(5, new ListNode(6, new ListNode(4))),
        ),
    ).toEqual(new ListNode(7, new ListNode(0, new ListNode(8))))
})

test('addTwoNumbers2', () => {
    expect(new ListNode(7).val).toEqual(7)
})

test('addTwoNumbers3', () => {
    expect(addTwoNumbers(new ListNode(5), new ListNode(5))).toEqual(new ListNode(0, new ListNode(1)))
})
