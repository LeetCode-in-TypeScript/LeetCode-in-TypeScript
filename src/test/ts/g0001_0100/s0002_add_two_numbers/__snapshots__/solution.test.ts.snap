// Vitest Snapshot v1

exports[`add_two_numbers > should be worked well 1`] = `
ListNode {
  "next": ListNode {
    "next": ListNode {
      "next": ListNode {
        "next": null,
        "val": 1,
      },
      "val": 3,
    },
    "val": 0,
  },
  "val": 2,
}
`;
